### ISSUE TITLE
Seeking in AES-128 Data Source doesn't work

### ISSUE THREAD

**Issue by StuffyEst** [google/ExoPlayer#2387]
Issue is that seeking does not work on default Aes128DataSource (ExoPlayer version exoplayer:r2.1.1)

I looked at default implementation of ExoPlayer Aes128DataSource. It does not support reading from arbitary position. So I implemented my own CustomAes128DataSource that build upon ExoPlayers private Aes128DataSource that has a random access input stream underneath. 

Everything works when playback is started with position 0. But as soon as I seek, player stops (position indicator stays put). It starts to work again when I seek back to 0.

If anyone has any idea what I'm missing, please comment.

Heres the code:

    @Override
    public long open(DataSpec dataSpec) throws IOException {
        this.encryptmode = dataSpec.uri.toString().endsWith(".ts");
        if (!this.encryptmode) {
            return this.upstream.open(dataSpec);
        }
        DataSourceInputStream encryptedInputStream = new DataSourceInputStream(upstream, dataSpec);

        Cipher cipher = createAESCipher(encryptionKey, encryptionIv);
        CipherInputStream cipherInputStream = new CipherInputStream(encryptedInputStream, cipher);
        byte[] decryptedByteArray = cipherInputStreamToByteArray(cipherInputStream);

        inputStream = new ByteArrayInputStream(decryptedByteArray);
        long skipped = inputStream.skip(dataSpec.position);
        if (skipped < dataSpec.position) {
            throw new EOFException();
        }

        if (dataSpec.length != C.LENGTH_UNSET) {
            bytesRemaining = dataSpec.length;
        } else {
            bytesRemaining = inputStream.available();
            if (bytesRemaining == 0) {
                bytesRemaining = C.LENGTH_UNSET;
            }
        }

        return bytesRemaining;
    }

    @Override
    public int read(byte[] buffer, int offset, int readLength) throws IOException {
        if (!this.encryptmode) {
            return this.upstream.read(buffer, offset, readLength);
        }
        
        if (bytesRemaining == 0) {
            return C.RESULT_END_OF_INPUT;
        } else {
            int bytesRead = 0;
            int bytesToRead = bytesRemaining == C.RESULT_END_OF_INPUT ? readLength : (int) Math.min(bytesRemaining, readLength);
            bytesRead = inputStream.read(buffer, offset, bytesToRead);
            if (bytesRead > 0) {
                if (bytesRemaining != C.RESULT_END_OF_INPUT) {
                    bytesRemaining -= bytesRead;
                }
            }
            return bytesRead;
        }
    }
`

**Comment by ojw28** [comment]
I'm not really sure what the issue you're facing actually is. `Aes128DataSource` is not part of the public API. It's package private to the HLS package and is not meant for use externally. Seeking specifically for HLS with AES-128 should work fine because the media is segmented into separate files.

Please clarify exactly what you're trying to do and why it's not working.

**Comment by StuffyEst** [comment]
The problem is that I have to feed in the encryptionKey and encryptionIV manually, because m3u8 file doesn't have #EXT-X-KEY.
So I took the package private Aes128DataSource and provided my own encryptionKey and encryptionIV to handle the decryption.

Everything works until I try to seek inside the SimpleExoPlayerView. The video output turns black, progress bar stops, play/pause button does nothing. When I keep the player at "Play" state, and seek back to position 0, the video starts to work again.

So I thought the problem was with reading data, and implemented my own data source because of that. Was that the wrong move and the error is somewhere else?

**Comment by ojw28** [comment]
It's not possible to say where the problem is because there's insufficient information provided (including how we can easily reproduce the issue or details of the changes you've made). There are no known issues with regular HLS with AES-128 playbacks. If you're inserting the key/IV and instantiating the source correctly then I wouldn't expect issues with your approach either.

**Comment by StuffyEst** [comment]
I could extract the logic to a sample android project and share it privatly, if you had the time to look at it?

**Comment by ojw28** [comment]
It's unlikely we'll have time to look into this unless you're able to share a (smallish) patch to the demo app that reproduces the issue, along with sample content to reproduce the problem. It sounds likely to be an application specific issue rather than a problem with ExoPlayer.

I can't think of a reason why you shouldn't share a patch publicly. You can send links to the media to `dev.exoplayer@gmail.com` if you'd prefer for those not to be public.

**Comment by StuffyEst** [comment]
Sent media url-s and very small android sample app to show the issue with default Aes128DataSource to dev.exoplayer@gmail.com
Android Studio 2.3 Beta 3, Android min API 19, compile API 25

**Comment by ojw28** [comment]
If you want us to look at this then please provide a small patch to our existing demo app, as opposed to a whole new demo app that we're not familiar with.

**Comment by StuffyEst** [comment]
I'm unable to even get the AES128 work in Demo application. I can't find a good place to swap out the DataSource to use. Is there a example to show how to add a custom datasource factory in Demo enviroment?

**Comment by StuffyEst** [comment]
I'll ask this in another way. I have a URL that points to .m3u8 with AES128 encryption. I have to provide the encryptionKey and encryptionIv manually. What is the correct approach to do so?

Edit: sent the closest representation that I could do to the demo app in a e-mail containing the patch. But this version throws 

```
01-30 02:48:20.513 3461-3461/com.google.android.exoplayer2.demo D/EventLogger: videoEnabled [203.12]
01-30 02:48:20.513 3461-3461/com.google.android.exoplayer2.demo D/EventLogger: audioEnabled [203.12]
01-30 02:48:20.521 3461-8262/com.google.android.exoplayer2.demo I/OMXClient: Using client-side OMX mux.
01-30 02:48:20.537 3461-8262/com.google.android.exoplayer2.demo E/OMXNodeInstance: OMX_GetExtensionIndex OMX.google.android.index.storeMetaDataInBuffers failed
01-30 02:48:20.537 3461-8262/com.google.android.exoplayer2.demo E/ACodec: [OMX.google.h264.decoder] storeMetaDataInBuffers failed w/ err -2147483648
01-30 02:48:20.539 3461-3461/com.google.android.exoplayer2.demo D/EventLogger: videoDecoderInitialized [203.15, OMX.google.h264.decoder]
01-30 02:48:20.539 3461-3461/com.google.android.exoplayer2.demo D/EventLogger: videoFormatChanged [203.15, id=null, mimeType=video/avc, res=640x360]
01-30 02:48:20.539 3461-8263/com.google.android.exoplayer2.demo E/SoftAVC: Decoder failed: -2
01-30 02:48:20.540 3461-8262/com.google.android.exoplayer2.demo E/ACodec: [OMX.google.h264.decoder] ERROR(0x80001001)
01-30 02:48:20.540 3461-8262/com.google.android.exoplayer2.demo E/ACodec: signalError(omxError 0x80001001, internalError -2147483648)
01-30 02:48:20.540 3461-8261/com.google.android.exoplayer2.demo E/MediaCodec: Codec reported err 0x80001001, actionCode 0, while in state 6
01-30 02:48:20.541 3461-5268/com.google.android.exoplayer2.demo E/ExoPlayerImplInternal: Internal runtime error.
```

**Comment by ojw28** [comment]
I took a look at this. I think your AES-128 logic is correct and working. It looks like the media (after removing the AES-128 encryption) violates this part of the HLS spec:

> Each Transport Stream Segment MUST contain a PAT and a PMT, ideally at the start of the Segment, or have an EXT-X-MAP (Section 4.3.2.5) tag applied to it.

Out of the segments that I looked at, only the first one contains the PAT/PMT. This is why seeking to 0 (or anywhere within the first segment, so up to 10s) works, whilst seeking after 10s does not.

If you remove the AES-128 encryption from all of the segments and try and play them individually (i.e. as regular TS files) I think you'll find that only the first segment plays (in any player) due to lack of PAT/PMT in the other segments. The solution here is to fix the media.
